Impl symmetric AES encryption on top of TcpStream

s1gtrap opened the pull request:
- The server will now generate a 1024-bit certificate and corresponding private key at startup which is used during the encryption handshake (the certificates public key is sent to the client after which the private key is used to decrypt the shared secret).
- Upon success the `TcpStream` is wrapped in the `SymmStream` type, after which all subsequent reads and writes are passed through two separate 128-bit AES ciphers (`decrypter` and `encrypter` respectively.)

The clients response is not verified atm, since the `login::clientbound::Disconnect` packet isn't implemented (guessing it won't be long, since @toqueteos [has made efforts towards encoding ChatJson?](https://github.com/toqueteos/hematite_server/commit/b3ee689d55e1d68bdbc44d224eaa18bcb2a85fc0))

The nonce isn't randomly generated either, which I suppose isn't strictly necessary, but would be nice to have. [It's also how cuberite does it (kind of.)](https://github.com/cuberite/cuberite/blob/83418e1d7a31e247b0d7892e75e70abc14b9aad8/src/Protocol/Protocol18x.cpp#L2205-L2206)

I'm also not happy with overruling the compiler in terms of `Send` and `Sync` safety (see [src/vanilla/server.rs](https://github.com/bheart/hematite_server/blob/b47d99ed3f6f891c149d4fbd9468af5b2d8618be/src/vanilla/server.rs#L144-L145)), but I'm under the impression that the `openssl` crate is as thread safe as it can get, and that marking `PKey`s as such is a non-issue (I could totally be wrong, I'm not particularly well versed in Rust.)

Still to do:
- [ ] Verify `verify_token`
- [ ] Generate `verify_token` as opposed to hardcoding it
- [x] Resolve thread safety issues (if any?)

s1gtrap committed:
Impl symmetric AES encryption on top of TcpStream

* The server generates a 1024-bit certificate and corresponding private
key at startup which is used during the encryption handshake.
* All connections are wrapped in the `SymmStream` type during login,
after which all r/w's are passed through separate 128-bit AES ciphers
(known as `Crypter`s in `openssl` crate.)

s1gtrap commented:
I'll get on this as soon as I can. `read_exact` can easily be replaced with `read_exactly` and `clone_from_slice`s implementation is also fairly straight forward.

toqueteos commented:
Welcom to hematite_server @bheart and thanks for sending in this PR!

You are right, `login::clientbound::Disconnect` is not implemented **on master**. It is implemented on ~~#97~~ #106 which hasn't been merged yet :sob:

Apparently [io::copy](https://doc.rust-lang.org/std/io/fn.copy.html) can be used with slices too. Would that help?

We probably should use `take` instead of `read_exactly` this is something I always forget to open an issue for...

toqueteos commented:
Oh, I forgot! It'll would be a good thing to enable encryption iff `online-mode` is `true` (#44).

s1gtrap commented:
Wasn't aware of `io::copy`! Thanks! The thing about `read_exact`/`read_exactly` is that they both error if they couldn't read the specified amount (right?), whereas an `io::Take` (when `read_to_end`) will happily return less. ~~Not sure if that's desirable tbh.~~

toqueteos commented:
Just to be sure, when you say `read_exactly` are you referencing the one that's implemented on `hematite_server/util`?

s1gtrap commented:
Yea, don't `read_exact` and `read_exactly` work in pretty much the same way?

Regardless, using an `io::Take` does make more sense, all things considered. There are valid reasons for why the bytes read can be less than the output buffer's size after all.

toqueteos commented:
@bheart Yeah, `read_exactly` is inspired/copied so we can use it on stable. `io::Take` makes sense but maybe for some specific cases we'll need the exact version.

toqueteos commented on src/world.rs:
```diff
@@ -68,7 +68,7 @@ impl World {
 
     #[allow(unreachable_code)]
-    pub fn handle_player(&self, mut stream: TcpStream) -> io::Result<()> {
+    pub fn handle_player(&self, mut stream: SymmStream) -> io::Result<()> {
```
We should change this to be a `S: Read + Write` instead of `TcpStream/SymmStream` so we can easily switch authenticated mode `on/off`. /cc @fenhl

s1gtrap replied:
That will be difficult to merge with your single-player branch, because of [the `try_clone` call](https://github.com/toqueteos/hematite_server/blob/cf110e63ef57be1049636e95dace6afd16b9980b/src/world.rs#L188).

I was thinking about introducing a `TryClone` trait and implementing it for the `SymmStream`, plus an additional `PlainStream` which would do nothing but wrap the underlying `TcpStream`.

So two structs, each with an inner `TcpStream`, one for plaintext communication and one for encrypted communication.

Then `S: Read + Write + TryClone` would work.

fenhl commented on src/vanilla/server.rs:
```diff
@@ -104,3 +140,6 @@ impl Server {
 }
+
+unsafe impl Send for Server {}
+unsafe impl Sync for Server {}
```
Undocumented unsafe code is absolutely unacceptable. Do not merge until this is fixed.

bheart replied:
Any ideas for dealing with this? The issue stems from the fact that `PKey` has a `*mut _` value. From what I can tell, the C type `EVP_KEY` is perfectly safe, provided a locking function is set (which the `openssl` crate handles nicely.)

s1gtrap replied:
Now that sfackler/rust-openssl@89e88727ffab25c49e67415ebc6bca555f6358c6 is in this isn't necessary anymore.

s1gtrap committed:
Use crates.io rust-openssl release

Uses rust-openssl as released on crates.io instead of pulling the
revision from Github.
